To convert and serve this notebook as a reveal.js slideshow:
```python
jupyter nbconvert SatPy.ipynb --to slides --post serve --SlidesExporter.reveal_theme=simple --SlidesExporter.reveal_scroll=True --SlidesExporter.reveal_transition=none
```

In [1]:
import warnings
warnings.filterwarnings('ignore')
import logging
logging.basicConfig(level=logging.CRITICAL)

# SatPy
## A Python Library for Weather Satellite Imagery

***

By David Hoese <sub>*pronounced Haze</sub>

# SatPy Core Developers

* Martin Raspaud
* Adam Dybbroe
* Panu Lahtinen

## Thank you

* PyTroll Users
* Kathy Strabala

# About Me

* Github: djhoese
* Twitter: djhoese
* Software Developer @ Space Science and Engineering Center (SSEC) - University of Wisconsin - Madison
* Community Satellite Processing Package (CSPP) Team
* Polar2Grid/Geo2Grid
* Satellite Information Familiarization Tool (SIFT) - Poster
* VisPy Maintainer - Plenary
* PyTroll/SatPy Maintainer

# Outline

1. What do we work with?
2. How does SatPy make our work easier?
3. Cool stuff you can do with SatPy

# What do we work with? Weather Satellite Data

* Forecasting and forecast models
* Research/Analysis
* Pretty pictures
* Operational processing
  
<img src="SatPy - Do Stuff Diagram.png" alt="Do Stuff Diagram">

In [1]:
from satpy import Scene
from glob import glob
scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
channels = ["C{channel:02d}".format(channel=chn) for chn in range(1, 17)]
scn.load(channels)
scn.save_datasets()
! montage C??_20170920_173040.tif -geometry 512x512+4+4  -background black montage_abi_20170920_173040.jpg

Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp


montage: Unknown field with tag 33550 (0x830e) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 33922 (0x8482) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34735 (0x87af) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34736 (0x87b0) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34737 (0x87b1) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 42112 (0xa480) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 33550 (0x830e) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 33922 (0x8482) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34735 (0x87af) encountered. `TIFFReadDirectory' 

<img src="montage_abi_20170920_173040.jpg" alt="ABI Montage" style="width:600px;">

In [15]:
from satpy import Scene
from glob import glob
scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['true_color'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
! gdal_translate -of PNG -outsize 10% 10% true_color_20170920_173040{.tif,.png}

Platform file /Users/davidh/repos/git/satpy/satpy/etc/platforms.txt not found.
Rasterio 1.0+ required for setting colorinterp
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in less
  return func(*args2)
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in greater_equal
  return func(*args2)
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in log10
  return func(*args2)


Input file size is 21696, 21696
0...10...20...30...40...50...60...70...80...90...100 - done.


<img src="true_color_20170920_173040.png" alt="ABI True Color" style="width:600px;">

# Our solution: SatPy

* Version 0.9 released
* Linux, OSX, Windows compatible
* Installable with ``pip`` and ``conda`` (conda-forge)

# How does SatPy make our work easier?

* Common interfaces
* Performance - ``xarray`` and ``dask``
* Community

# Common Interfaces

* Read
* Composite/Modify
* Resample
* Write

# Read: How do I get the information I need?

* Reading satellite data is annoying
  * Formats: NetCDF, HDF5, custom binary, compressed
  * File "schemes" and organization
  * Fill values
  * Concatenation
  * Missing metadata
  * Many data variations
  * Standards

<img src="https://imgs.xkcd.com/comics/standards.png" alt="drawing">

In [2]:
from satpy import available_readers
sorted(available_readers())

['abi_l1b',
 'acspo',
 'ahi_hsd',
 'amsr2_l1b',
 'avhrr_aapp_l1b',
 'avhrr_eps_l1b',
 'clavrx',
 'fci_fdhsi',
 'generic_image',
 'geocat',
 'ghrsst_osisaf',
 'grib',
 'hdf4_caliopv3',
 'hdfeos_l1b',
 'hrit_electrol',
 'hrit_goes',
 'hrit_jma',
 'hrit_msg',
 'iasi_l2',
 'li_l2',
 'maia',
 'native_msg',
 'nc_nwcsaf_msg',
 'nc_nwcsaf_pps',
 'nc_olci_l1b',
 'nc_olci_l2',
 'nc_slstr',
 'nucaps',
 'omps_edr',
 'safe_sar_c',
 'scatsat1_l2b',
 'viirs_compact',
 'viirs_l1b',
 'viirs_sdr']

# Loading data and the Scene object

* By Name
* By Wavelength
* By ``DatasetID``
* Specifying variations
  * resolution
  * polarization
  * calibration - counts, radiance, reflectance, brightness temperature, etc.

In [32]:
from satpy import Scene, DatasetID
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['C01', 'C02'])

In [33]:
scn.load([0.46, 0.62])

In [34]:
ds_id = DatasetID(name='C02', calibration='radiance')
scn.load([ds_id])

In [26]:
scn.available_dataset_names()

['C01', 'C02', 'C03', 'C04', 'C05', 'C06', 'C07', 'C08', 'C09', 'C10', 'C11', 'C12', 'C13', 'C14', 'C15', 'C16']

In [30]:
scn.available_dataset_ids()

[DatasetID(name='C01', wavelength=(0.45, 0.47, 0.49), resolution=1000, polarization=None, calibration='radiance', level=None, modifiers=()),
 DatasetID(name='C01', wavelength=(0.45, 0.47, 0.49), resolution=1000, polarization=None, calibration='reflectance', level=None, modifiers=()),
 DatasetID(name='C02', wavelength=(0.59, 0.64, 0.69), resolution=500, polarization=None, calibration='radiance', level=None, modifiers=()),
 DatasetID(name='C02', wavelength=(0.59, 0.64, 0.69), resolution=500, polarization=None, calibration='reflectance', level=None, modifiers=()),
 DatasetID(name='C03', wavelength=(0.8455, 0.865, 0.8845), resolution=1000, polarization=None, calibration='radiance', level=None, modifiers=()),
 DatasetID(name='C03', wavelength=(0.8455, 0.865, 0.8845), resolution=1000, polarization=None, calibration='reflectance', level=None, modifiers=()),
 DatasetID(name='C04', wavelength=(1.3705, 1.378, 1.3855), resolution=2000, polarization=None, calibration='radiance', level=None, modifi

In [29]:
scn['C02']

<xarray.DataArray (y: 21696, x: 21696)>
dask.array<shape=(21696, 21696), dtype=float64, chunksize=(4096, 4096)>
Coordinates:
  * x        (x) float64 -0.1519 -0.1519 -0.1518 -0.1518 -0.1518 -0.1518 ...
    y_image  float32 0.0
    x_image  float32 0.0
  * y        (y) float64 0.1519 0.1519 0.1518 0.1518 0.1518 0.1518 0.1518 ...
Attributes:
    satellite_longitude:    -89.5
    satellite_latitude:     0.0
    satellite_altitude:     35786.0234375
    units:                  %
    modifiers:              ()
    wavelength:             (0.59, 0.64, 0.69)
    grid_mapping:           goes_imager_projection
    _Unsigned:              true
    cell_methods:           t: point area: point
    standard_name:          toa_bidirectional_reflectance
    scale_factor:           0.158592
    ancillary_variables:    []
    valid_range:            [   0 4094]
    long_name:              ABI L1b Radiances
    add_offset:             -20.2899
    resolution:             500
    calibration:            

In [14]:
scn.show('C01')

In [4]:
scn.save_dataset('C01', base_dir='/Users/davidh/repos/git/pytroll-examples/satpy/presentations/2018_scipy/', writer='simple_image')

! gdal_translate -of PNG -outsize 10% 10% C01_20170920_173040.png C01_20170920_173040.small.png

Input file size is 10848, 10848
0...10...20...30...40...50...60...70...80...90...100 - done.


<img src="C01_20170920_173040.small.png" alt="ABI C01" style="width:550px;">

# Composite: How do I make pretty color pictures?

<img src="abi_rgbs_20170920_173040.jpg" alt="ABI RGB montage" style="width:550px;">

In [ ]:
from satpy import Scene
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['overview', 'natural', 'airmass'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
! montage {true_color,overview,natural,airmass}_20170920_173040.tif -geometry 512x512+2+2 -background black abi_rgbs_20170920_173040.jpg

Platform file /Users/davidh/repos/git/satpy/satpy/etc/platforms.txt not found.
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in less
  return func(*args2)
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in greater_equal
  return func(*args2)


In [1]:
from satpy import Scene
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.available_composite_names()

['airmass', 'ash', 'dust', 'fog', 'green', 'green_raw', 'green_snow', 'ir_cloud_day', 'natural', 'natural_raw', 'natural_sun', 'night_microphysics', 'overview', 'overview_raw', 'true_color', 'true_color_raw']

In [2]:
scn.load(['airmass'])
scn['airmass']

<xarray.DataArray 'concatenate-d3ffb434ba4f6164ba94e209a0860c74' (bands: 3, y: 5424, x: 5424)>
dask.array<shape=(3, 5424, 5424), dtype=float64, chunksize=(1, 4096, 4096)>
Coordinates:
  * x        (x) float64 -0.1518 -0.1518 -0.1517 -0.1517 -0.1516 -0.1516 ...
    y_image  float32 0.0
    x_image  float32 0.0
  * y        (y) float64 0.1518 0.1518 0.1517 0.1517 0.1516 0.1516 0.1515 ...
  * bands    (bands) <U1 'R' 'G' 'B'
Attributes:
    long_name:               ABI L1b Radiances
    grid_mapping:            goes_imager_projection
    level:                   None
    satellite_latitude:      0.0
    sensor:                  abi
    satellite_altitude:      35786.0234375
    ancillary_variables:     []
    start_time:              2017-09-20 17:30:40.800000
    cell_methods:            t: point area: point
    polarization:            None
    resolution:              None
    area:                    Area ID: some_area_name\nDescription: On-the-fly...
    platform_name:           GOES

# Composite YAML configuration files

```yaml
  airmass:
    compositor: !!python/name:satpy.composites.Airmass
    prerequisites:
    - 6.2
    - 7.3
    - 9.7
    - 10.3
    standard_name: airmass
```

# Modifiers: How do I make my pretty picture prettier?

<img src="true_color_before_after_20170920_173040.jpg" alt="ABI True Color Before/After">

# Do the impossible

In [ ]:
scn.load(['true_color'])

Operations Involved:

* 3 input bands (500m and 1000m)
* 3 complex data enhancements
* 1 pseudo-band (green)
* Upscale to 500m
* Final result: 21696 rows \* 21696 cols \* 8 bytes \* 3 input bands = ~10.5 GB
* Dask-based SatPy loads, computes, and saves to geotiff in 6-8 minutes

In [1]:
from satpy import Scene
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['true_color_raw'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
! montage true_color_raw_20170920_173040.tif true_color_20170920_173040.tif -geometry 512x512+2 -background black true_color_before_after_20170920_173040.jpg

# Modifiers

* During load

```python
ds_id = DatasetID(name='C01', modifiers=('sunz_corrected',))
scn.load([ds_id])
```

* Configuring composites

```yaml

  true_color:
    compositor: !!python/name:satpy.composites.SelfSharpenedRGB
    prerequisites:
    - name: C02
      modifiers: [sunz_corrected, rayleigh_corrected]
    - name: green
    - name: C01
      modifiers: [sunz_corrected, rayleigh_corrected]
    standard_name: true_color

```

# Resampling

* Remap to a common projection/grid
* Change projections

* Users want a common projection independent of the polar orbit
* Resampling in order to remap to one common projection


In [36]:
from satpy import Scene
from glob import glob
scn = Scene(reader='viirs_sdr',
            filenames=glob('/data/data/viirs/conus_day/*t1801*.h5'))
scn.load(['I04'])

In [ ]:
scn.show('I04')

<img src="I04_20120225_180125.swath.png" alt="VIIRS I04 Swath">

In [ ]:
scn = Scene(reader='viirs_sdr',
            filenames=glob('/data/data/viirs/conus_day/*t180*.h5'))
scn.load(['I04'])
new_scn = scn.resample('211e')
new_scn.show('I04')

<img src="I04_20120225_180125.png" alt="NPP VIIRS I04" style="width:450px;">

In [ ]:
from pyresample.geometry import AreaDefinition
area_def = AreaDefinition(
    'name', 'description', 'merc',
    {'proj': 'merc', 'datum': 'WGS84', 'lon_0': -95.0},
    1000, 1000, (0, 1750000, 4500000, 5500000))

new_scn = scn.resample(area_def, radius_of_influence=10000)
new_scn.show('I04')

<img src="I04_20120225_180125.custom_area.png" alt="NPP VIIRS I04 Custom" style="width:450px;">

# Saving to disk

* Writers
  * Geotiff (rasterio)
  * Pillow (png, jpeg, etc)
  * AWIPS SCMI
  * CF-compliant NetCDF (limited projection support)
* Use dask blocks when possible

In [63]:
from satpy import available_writers
available_writers()

['mitiff', 'simple_image', 'geotiff', 'cf', 'scmi']

In [ ]:
scn.save_datasets(writer='geotiff', base_dir='/tmp', compress='LZW')

# The Basics

```python
from satpy import Scene
scn = Scene(reader='abi_l1b', filenames=[...])
scn.load(['C01', 'C02', 'C08', 'true_color', 'air_mass', 'natural'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets(base_dir='/tmp')
```

# Fun Stuff: Stacked Scenes

In [2]:
from satpy import Scene, MultiScene
from glob import glob
import os
from dask.diagnostics import ProgressBar
base_dir = '/data/data/viirs/2018_06_29/'
viirs_passes = ['2018_06_29_180_0614', '2018_06_29_180_0752', '2018_06_29_180_0934']
viirs_scenes = [Scene(reader='viirs_sdr',
                      filenames=glob(os.path.join(base_dir, vp, '*.h5'))) 
                for vp in viirs_passes]

mscn = MultiScene(viirs_scenes)
mscn.load(['I04'])
new_mscn = mscn.resample('211e')
blended_scn = new_mscn.blend()
with ProgressBar():
    blended_scn.save_datasets()

[########################################] | 100% Completed |  4min 15.4s
[########################################] | 100% Completed |  0.1s


In [4]:
! gdal_translate -of PNG -outsize 10% 10% I04_20180629_061415{.tif,.png}

Input file size is 5120, 5120
0...10...20...30...40...50...60...70...80...90...100 - done.


<img src="I04_20180629_061415.png" alt="Blended I04" style="width:550px">

# Fun Stuff: Animations

```python
scenes = [Scene(reader='abi_l1b', filenames=step_files)
          for step_files in grouped_files]
mscn = MultiScene(scenes)
mscn.load(['C01'])
new_mscn = mscn.resample(resampler='native')
new_mscn.save_animation('{name}_{start_time:%Y%m%d_%H%M%S}.mp4',
                        fps=5, batch_size=4,
                        output_params=["-vf", "format=yuv420p,scale=640:-1"])
```

In [5]:
from IPython.display import HTML
HTML("""
<video width="640" height="480" controls>
  <source src="C01_20180623_000045.mp4" type="video/mp4">
</video>
""")

# Fun Stuff: Cartopy

In [ ]:
from satpy import Scene
from glob import glob
import matplotlib.pyplot as plt

filenames = glob('/data/data/viirs/conus_day/*t180*.h5')
scn = Scene(reader='viirs_sdr', filenames=filenames)
scn.load(['I04'])
new_scn = scn.resample('211e')

crs = new_scn['I04'].attrs['area'].to_cartopy_crs()
ax = plt.axes(projection=crs)

ax.coastlines()
ax.gridlines()
ax.set_global()
plt.imshow(new_scn['I04'], transform=crs, extent=crs.bounds, origin='upper')
cbar = plt.colorbar()
cbar.set_label("Kelvin")
plt.show()

<img src="I04_cartopy_plot.png" alt="I04 Cartopy">

# SatPy and The PyTroll Community

* Community Website: http://pytroll.github.io/
* GitHub: https://github.com/pytroll/satpy
* SatPy Documentation: http://satpy.readthedocs.io/en/latest/
* Mailing List: https://groups.google.com/d/forum/pytroll
* Workshops twice a year!
* Chat with us on slack (invite link on community website)

<img src="https://raw.githubusercontent.com/pytroll/pytroll/master/web/source/images/pytroll_dark_small.png" style="background-color:black;">


In [5]:
from IPython.display import IFrame
IFrame('https://www.youtube.com/embed/eBQi2G_fqXQ?rel=0', width=600, height=400)